### PR TITLE
update indexstorage interface to reduce roundtrips

### PR DESCRIPTION
**Pull request by bobcallaway**
This changes the interface for `indexstorage` in two ways:

1. Adds a `Shutdown()` method to clean up resources upon server shutdown
2. Takes `[]string` on both read and write paths

The redis implementation now pipelines requests into a single roundtrip between Rekor and Redis on both the read and write path (where possible).

### CODE REVIEW

**Review by cpanato — DISMISSED** [review]
lgtm

File: pkg/api/index.go [hunk 1]
```diff
@@ -42,14 +42,20 @@ func SearchIndexHandler(params index.SearchIndexParams) middleware.Responder {
...
+	var lookupKeys []string
+
 	if params.Query.Hash != "" {
 		// This must be a valid hash
-		sha := util.PrefixSHA(params.Query.Hash)
-		resultUUIDs, err := indexStorageClient.LookupIndices(httpReqCtx, strings.ToLower(sha))
-		if err != nil {
-			return handleRekorAPIError(params, http.StatusInternalServerError, fmt.Errorf("index storage error: %w", err), indexStorageUnexpectedResult)
+		sha := strings.ToLower(util.PrefixSHA(params.Query.Hash))
+		if queryOperator == "or" {
```
Comment (Hayden-IO) [inline pkg/api/index.go]: Does "and" need to be handled too? I think the lookups would be the same for or vs and, it seems like NewCollection just processes them different I think

File: pkg/indexstorage/redis/redis_test.go [hunk 2]
```diff
@@ -26,14 +26,14 @@ import (
 )
```
Comment (Hayden-IO) [inline pkg/indexstorage/redis/redis_test.go]: Can we add a test to confirm read/write works with multiple values?

File: pkg/api/index.go (hunk 1 above)
Comment (bobcallaway) [reply pkg/api/index.go]: It's handled in the else clause on line 52. The way the Collector was implemented is that you have to pass it in the queries for each input parameter separately, where as OR is just a concatenation.

File: pkg/indexstorage/redis/redis_test.go (hunk 2 above)
Comment (bobcallaway) [reply pkg/indexstorage/redis/redis_test.go]: The e2e tests already cover this

**Review by Hayden-IO — DISMISSED** [review]

**Review by Hayden-IO — APPROVED** [review]

**Pull request merged**